thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,163 @@
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include "slave.hpp"
+#include "jqueue.tpp"
+#include "jthread.hpp"
+#include "limits.h"
+
+
+class CTracingThorDataLink :  implements CInterfaceOf<IThorDataLink>
+{
+private:
+    class TraceQueue
+    {
+    private:
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            // NOTE - we don't output every item in the list to be 
+            // consistent with the other dump() implementation
+            // Note - whilst dumping, does not clear ring buffer
+            // so that it is available later
+            unsigned rowPos = rowHeadp;
+            for (unsigned n=1; n<traceRowBufferSize;++n)
+            {
+                if (++rowPos==traceRowBufferSize) rowPos = 0;  // NOTE - skipping first item intentional
+                OwnedConstThorRow row (rowBuffer[rowPos]);
+                if (!row) continue;
+
+                CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+                helper->queryOutputMeta()->toXML((const byte*)(row.get()), xmlwrite);
+                VStringBuffer trace("<Row>%s</Row>", xmlwrite.str());
+                mb.append(trace.str());
+            }
+        }
+    public:
+        OwnedConstThorRow *rowBuffer;
+        unsigned rowHeadp;
+        TraceQueue()
+        {
+            rowBuffer = NULL;
+            rowHeadp = 0;
+        }
+        ~TraceQueue()
+        {
+            delete [] rowBuffer;
+        }
+        void init()
+        {
+            if (!rowBuffer)
+            {
+                rowBuffer = new OwnedConstThorRow[traceRowBufferSize];
+                rowHeadp = 0;
+            }
+        }
+        void enqueue(const void *row)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowHeadp].set(row);
+            unsigned nextHead = rowHeadp + 1;
+            if (nextHead==traceRowBufferSize)
+                nextHead = 0;
+            rowHeadp = nextHead;
+        }
+        // This dump() makes use of pass-through buffer to ensure if it is called again
+        // then previous rows still available
+        void dump(MemoryBuffer &mb, TraceQueue &rowLogCache, IHThorArg *helper)
+        {
+            // NOTE - we don't output every item in the list as one
+            // COULD have been updated after we swapped the buffers
```

Comment:
Also (and maybe we discussed this(?), it does mean that if there's that race condition, that rows will appear out of sequence. e.g.:
+ buffers[0] = [a],[b],[*c] (c about to be written)
+ dump swaps, a,b added to tracing, buffers[1] now active
+ buffers[1] = [d],[e],[*f] (f about to be written)
+ dump swap, d,e added to tracing, buffers[0] now active
+ buffers[0] = [],[],[c],[g],[h],[*i] (i about to be written)
+ dump swap, c,g,h added to tracing, buffers[1] now active
+ buffers[1] = [],[],[f],[j],[k],[*l] (l about to be written)
+ dump swap, f,j,k added to tracing..


Should we just accept that, and just put a comment to mention that this can happen?

Or could we look at the other buffer (the one about to be swapped in) for the trailing row and add to the tracing circular ring, before swapping in ?
I think that could mean there's still a window if 2 swaps happened very quickly, where it would still be active and not picked up and therefore still lead to out order rows, but in practice with the time taken between 2 swaps over 2 debugRequest() hops that isn't going to happen.